src/hotspot/share/prims/jvmtiRedefineClasses.hpp

```diff
@@ -454,51 +454,46 @@ class VM_RedefineClasses: public VM_Operation {
     InstanceKlass* scratch_class, TRAPS);
   u2 rewrite_cp_ref_in_annotation_data(
     AnnotationArray* annotations_typeArray, int &byte_i_ref,
-    const char * trace_mesg, TRAPS);
+    const char * trace_mesg);
   bool rewrite_cp_refs(InstanceKlass* scratch_class, TRAPS);
   bool rewrite_cp_refs_in_annotation_struct(
-    AnnotationArray* class_annotations, int &byte_i_ref, TRAPS);
+    AnnotationArray* class_annotations, int &byte_i_ref);
   bool rewrite_cp_refs_in_annotations_typeArray(
-    AnnotationArray* annotations_typeArray, int &byte_i_ref, TRAPS);
-  bool rewrite_cp_refs_in_class_annotations(
-    InstanceKlass* scratch_class, TRAPS);
+    AnnotationArray* annotations_typeArray, int &byte_i_ref);
+  bool rewrite_cp_refs_in_class_annotations(InstanceKlass* scratch_class);
   bool rewrite_cp_refs_in_element_value(
-    AnnotationArray* class_annotations, int &byte_i_ref, TRAPS);
+    AnnotationArray* class_annotations, int &byte_i_ref);
   bool rewrite_cp_refs_in_type_annotations_typeArray(
     AnnotationArray* type_annotations_typeArray, int &byte_i_ref,
-    const char * location_mesg, TRAPS);
+    const char * location_mesg);
   bool rewrite_cp_refs_in_type_annotation_struct(
     AnnotationArray* type_annotations_typeArray, int &byte_i_ref,
-    const char * location_mesg, TRAPS);
+    const char * location_mesg);
   bool skip_type_annotation_target(
     AnnotationArray* type_annotations_typeArray, int &byte_i_ref,
-    const char * location_mesg, TRAPS);
+    const char * location_mesg);
   bool skip_type_annotation_type_path(
-    AnnotationArray* type_annotations_typeArray, int &byte_i_ref, TRAPS);
-  bool rewrite_cp_refs_in_fields_annotations(
-    InstanceKlass* scratch_class, TRAPS);
+    AnnotationArray* type_annotations_typeArray, int &byte_i_ref);
+  bool rewrite_cp_refs_in_fields_annotations(InstanceKlass* scratch_class);
   bool rewrite_cp_refs_in_nest_attributes(InstanceKlass* scratch_class);
-  bool rewrite_cp_refs_in_record_attribute(InstanceKlass* scratch_class, TRAPS);
+  bool rewrite_cp_refs_in_record_attribute(InstanceKlass* scratch_class);
   bool rewrite_cp_refs_in_permitted_subclasses_attribute(InstanceKlass* scratch_class);
+
   void rewrite_cp_refs_in_method(methodHandle method,
     methodHandle * new_method_p, TRAPS);
   bool rewrite_cp_refs_in_methods(InstanceKlass* scratch_class, TRAPS);
```

Comment:
`VM_RedefineClasses::load_new_class_versions` also seems to never throw. These functions should be changed to take a `Thread*` parameter, and should use `HandleMark em(thread);` to guarantee that an exception never leaves the function.

Reply:
Both of these functions are good examples of the convention that we're trying to agree on.   In, load_new_class_versions, TRAPS is passed to materialize THREAD.  THREAD is used then in a lot of places, and also to pass to SystemDictionary::parse_stream(...TRAPS), which does have an exceptional return that must be handled.
Removing TRAPS then adding:
JavaThread* current = JavaThread::current();
changing THREAD to current in most of the places seems ok, but passing 'current' to SystemDictionary::resolve_from_stream loses the information visually that this function returns an exception that must be handled.
We need some meta-writeup rather than these decisions made in pull requests, because if we made these decisions collectively, I missed out.